tools/libSwiftScan/libSwiftScan.cpp

```diff
@@ -777,15 +882,618 @@ swiftscan_cache_compute_key(swiftscan_cas_t cas, int argc, const char **argv,
+    return nullptr;
+  }
+
+  return wrap(Instance);
+}
+
+void swiftscan_cache_replay_instance_dispose(
+    swiftscan_cache_replay_instance_t instance) {
+  delete unwrap(instance);
+}
+
+namespace {
+class StreamOutputFileImpl final
+    : public llvm::RTTIExtends<StreamOutputFileImpl,
+                               llvm::vfs::OutputFileImpl> {
+
+public:
+  StreamOutputFileImpl(llvm::raw_pwrite_stream &OS) : OS(OS) {}
+  llvm::Error keep() final { return llvm::Error::success(); }
+  llvm::Error discard() final { return llvm::Error::success(); }
+  llvm::raw_pwrite_stream &getOS() final { return OS; }
+
+private:
+  llvm::raw_pwrite_stream &OS;
+};
+
+// The replay output backend. Currently, it redirects "-" to the
+// stdout provided.
+struct ReplayOutputBackend : public llvm::vfs::ProxyOutputBackend {
+  llvm::Expected<std::unique_ptr<llvm::vfs::OutputFileImpl>>
+  createFileImpl(llvm::StringRef Path,
+                 std::optional<llvm::vfs::OutputConfig> Config) override {
+    if (Path == "-")
+      return std::make_unique<StreamOutputFileImpl>(StdOut);
+    return llvm::vfs::ProxyOutputBackend::createFileImpl(Path, Config);
+  }
+
+  llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend>
+  cloneImpl() const override {
+    return llvm::makeIntrusiveRefCnt<ReplayOutputBackend>(
+        getUnderlyingBackend().clone(), StdOut);
+  }
+
+  ReplayOutputBackend(
+      llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> UnderlyingBackend,
+      llvm::raw_pwrite_stream &StdOut)
+      : ProxyOutputBackend(UnderlyingBackend), StdOut(StdOut) {}
+
+  llvm::raw_pwrite_stream &StdOut;
+};
+
+llvm::Expected<bool> replayCompilation(SwiftScanReplayInstance &Instance,
+                                       llvm::cas::ObjectStore &CAS,
+                                       llvm::cas::ActionCache &Cache,
+                                       SwiftCachedCompilation &Comp,
+                                       llvm::raw_pwrite_stream &Out,
+                                       llvm::raw_pwrite_stream &Err) {
+  using namespace swift;
+  using namespace llvm;
+  CompilerInstance Inst;
+  CompilerInvocation &Invocation = Instance.Invocation;
+
+  // Find the input file from the invocation.
+  auto &InputsAndOutputs =
+      Instance.Invocation.getFrontendOptions().InputsAndOutputs;
+  auto AllInputs = InputsAndOutputs.getAllInputs();
+  auto Input = llvm::find_if(AllInputs, [&](const InputFile &Input) {
+    return Input.getFileName() == Comp.CorrespondingInput;
+  });
+  if (Input == AllInputs.end())
+    return createStringError(inconvertibleErrorCode(),
+                             "InputFile \"" + Comp.CorrespondingInput +
+                                 "\" is not part of the compilation");
+
+  // Setup DiagnosticsConsumers.
+  // FIXME: Need to add support for JSONFIXIT, SerializedDiagnostics, etc.
+  PrintingDiagnosticConsumer PDC(Err);
+  Inst.addDiagnosticConsumer(&PDC);
+
+  if (Invocation.getDiagnosticOptions().UseColor)
```

Comment:
I guess in the long run it would be great to find a balance of how much of this we can factor out to something common as used by `performFrontend`.

Reply:
added FIXME